### Загрузим датасет

In [2]:
from sklearn.datasets import load_diabetes

Выведем информацию о датасете

In [4]:
print(load_diabetes().DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

Набор данных содержит:

-    Набор данных о диабете: Набор данных пациентов с диабетом.
-    Количество экземпляров: 442 пациента.
-    Количество атрибутов: Первые 10 столбцов являются числовыми предиктивными значениями.
-    Целевая переменная: Столбец 11 является количественной мерой прогрессирования заболевания через год после исходного уровня.
-    Информация об атрибутах:
       - __age__: возраст в годах
       - __sex__: пол
       - __bmi__: индекс массы тела
       - __bp__: среднее артериальное давление
       - __s1__: tc, общий сывороточный холестерин
       - __s2__: ldl, липопротеины низкой плотности

По сути, набор данных содержит информацию о 442 пациентах с диабетом, включая десять базовых переменных (возраст, пол, индекс массы тела, среднее артериальное давление и шесть измерений сыворотки крови), а также количественную меру прогрессирования заболевания через год после базового уровня.

In [23]:
df = load_diabetes(as_frame=True).frame

In [24]:
df.head()

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  target  
0 -0.002592  0.019907 -0.017646   151.0  
1 -0.039493 -0.068332 -0.092204    75.0  
2 -0.002592  0.002861 -0.025930   141.0  
3  0.034309  0.022688 -0.009362   206.0  
4 -0.002592 -0.031988 -0.046641   135.0

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442 entries, 0 to 441
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     442 non-null    float64
 1   sex     442 non-null    float64
 2   bmi     442 non-null    float64
 3   bp      442 non-null    float64
 4   s1      442 non-null    float64
 5   s2      442 non-null    float64
 6   s3      442 non-null    float64
 7   s4      442 non-null    float64
 8   s5      442 non-null    float64
 9   s6      442 non-null    float64
 10  target  442 non-null    float64
dtypes: float64(11)
memory usage: 38.1 KB


Пропусков нет, все данные имеют тип float64, размер датасета - 442

Подготовим данные к обучению моделей: отделите целевой признак от датасета, разделите датасет на тренировочную и валидационную части.

In [27]:
from sklearn.model_selection import train_test_split

X = df.drop(['target'], axis=1)
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.2, random_state=42)

Создадим и обучим нашу модель

In [29]:
from sklearn.tree import DecisionTreeRegressor

model_tree = DecisionTreeRegressor()
model_tree.fit(X_train, y_train)

y_pred_tree = model_tree.predict(X_test)

Оценим адекватность обученной модели

In [37]:
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

mse_tree = mean_squared_error(y_test, y_pred_tree)
rmse_tree = np.sqrt(mse_tree)
r2_tree = r2_score(y_test, y_pred_tree)

print(f'MSE для DecisionTreeRegressor: {mse_tree:.2f}')
print(f'RMSE для DecisionTreeRegressor: {rmse_tree:.2f}')
print(f'R2 для DecisionTreeRegressor: {r2_tree:.2f}')

MSE для DecisionTreeRegressor: 6516.98
RMSE для DecisionTreeRegressor: 80.73
R2 для DecisionTreeRegressor: -0.09


In [38]:
y_pred_basline = np.full_like(y_test, y_train.mean())

mse_basline = mean_squared_error(y_test, y_pred_basline)
rmse_basline = np.sqrt(mse_basline)
r2_basline = r2_score(y_test, y_pred_basline)

print(f'MSE для простой модели: {mse_basline:.2f}')
print(f'RMSE для простой модели: {rmse_basline:.2f}')
print(f'R2 для простой модели: {r2_basline:.2f}')

MSE для простой модели: 6204.35
RMSE для простой модели: 78.77
R2 для простой модели: -0.04


Простая модель с прогнозом справилась лучше

In [39]:
from sklearn.linear_model import LinearRegression

model_linear = LinearRegression()
model_linear.fit(X_train, y_train)

y_pred_linear = model_linear.predict(X_test)

mse_linear = mean_squared_error(y_test, y_pred_linear)
rmse_linear = np.sqrt(mse_basline)
r2_linear = r2_score(y_test, y_pred_linear)

print(f'MSE для LinearRegression: {mse_linear:.2f}')
print(f'RMSE для LinearRegression: {rmse_linear:.2f}')
print(f'R2 для LinearRegression: {r2_linear:.2f}')

MSE для LinearRegression: 3371.17
RMSE для LinearRegression: 78.77
R2 для LinearRegression: 0.44


Модель линейной регрессии справилась с прогнозом лучше простой и модели и DecisionTreeRegressor